// @aws-solutions-constructs/core/lib/fargate-defaults.js
"use strict";
/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultContainerDefinitionProps = exports.DefaultApplicationTargetGroupProps = exports.DefaultFargateTaskDefinitionProps = exports.DefaultFargateServiceProps = exports.DefaultClusterProps = void 0;
const ecs = require("aws-cdk-lib/aws-ecs");
const elb = require("aws-cdk-lib/aws-elasticloadbalancingv2");
function DefaultClusterProps() {
    return {};
}
exports.DefaultClusterProps = DefaultClusterProps;
function DefaultFargateServiceProps() {
    return {
        assignPublicIp: false,
        desiredCount: 2,
        maxHealthyPercent: 150,
        minHealthyPercent: 75,
        platformVersion: ecs.FargatePlatformVersion.LATEST,
    };
}
exports.DefaultFargateServiceProps = DefaultFargateServiceProps;
function DefaultFargateTaskDefinitionProps() {
    return {
        cpu: 256,
        memoryLimitMiB: 512,
    };
}
exports.DefaultFargateTaskDefinitionProps = DefaultFargateTaskDefinitionProps;
function DefaultApplicationTargetGroupProps(vpc) {
    return {
        port: 8080,
        protocol: elb.ApplicationProtocol.HTTP,
        protocolVersion: elb.ApplicationProtocolVersion.HTTP1,
        targetType: elb.TargetType.IP,
        vpc,
        healthCheck: { path: '/' }
    };
}
exports.DefaultApplicationTargetGroupProps = DefaultApplicationTargetGroupProps;
function DefaultContainerDefinitionProps() {
    return {
        memoryReservationMiB: 512,
        portMappings: [{ containerPort: 8080 }],
    };
}
exports.DefaultContainerDefinitionProps = DefaultContainerDefinitionProps;
