// @aws-solutions-constructs/core/lib/security-group-helper.js
"use strict";
/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateSelfReferencingSecurityGroup = exports.buildSecurityGroup = void 0;
// import { countResources } from "@aws-cdk/assert";
const ec2 = require("aws-cdk-lib/aws-ec2");
const utils_1 = require("./utils");
function buildSecurityGroup(scope, name, props, ingressRules, egressRules) {
    const newSecurityGroup = new ec2.SecurityGroup(scope, `${name}-security-group`, props);
    ingressRules.forEach(rule => {
        newSecurityGroup.addIngressRule(rule.peer, rule.connection, rule.description, rule.remoteRule);
    });
    egressRules.forEach(rule => {
        newSecurityGroup.addEgressRule(rule.peer, rule.connection, rule.description, rule.remoteRule);
    });
    utils_1.addCfnSuppressRules(newSecurityGroup, [
        {
            id: "W5",
            reason: "Egress of 0.0.0.0/0 is default and generally considered OK",
        },
        {
            id: "W40",
            reason: "Egress IPProtocol of -1 is default and generally considered OK",
        },
    ]);
    return newSecurityGroup;
}
exports.buildSecurityGroup = buildSecurityGroup;
function CreateSelfReferencingSecurityGroup(scope, id, vpc, cachePort) {
    const newCacheSG = new ec2.SecurityGroup(scope, `${id}-cachesg`, {
        vpc,
        allowAllOutbound: true,
    });
    const selfReferenceRule = new ec2.CfnSecurityGroupIngress(scope, `${id}-ingress`, {
        groupId: newCacheSG.securityGroupId,
        sourceSecurityGroupId: newCacheSG.securityGroupId,
        ipProtocol: "TCP",
        fromPort: cachePort,
        toPort: cachePort,
        description: 'Self referencing rule to control access to Elasticache memcached cluster',
    });
    selfReferenceRule.node.addDependency(newCacheSG);
    utils_1.addCfnSuppressRules(newCacheSG, [
        {
            id: "W5",
            reason: "Egress of 0.0.0.0/0 is default and generally considered OK",
        },
        {
            id: "W40",
            reason: "Egress IPProtocol of -1 is default and generally considered OK",
        },
    ]);
    return newCacheSG;
}
exports.CreateSelfReferencingSecurityGroup = CreateSelfReferencingSecurityGroup;
